project1/server/agent.cpp

```diff
@@ -4,231 +4,415 @@
+            // the size.
+            incoming_message_buffer_.resize(bytes_read);
+
+            // Add the data to the parser.
+            parser_.AddNewData(incoming_message_buffer_);
+        }
+
+        // Get the parsed message.
+        if (!parser_.GetMessage(message)) {
+            std::cerr << "Failed to get the parsed message." << std::endl;
+            return ClientState::ERROR;
+        }
+        return ClientState::ACTIVE;
+    }
+
+    Agent::ClientState Agent::DispatchMessage(const Request &message) {
+        if (message.has_get()) {
+            return HandleRequest(message.get());
+        } else if (message.has_put()) {
+            return HandleRequest(message.put());
+        } else if (message.has_delete_()) {
+            return HandleRequest(message.delete_());
+        } else if (message.has_list()) {
+            return HandleRequest(message.list());
+        } else if (message.has_change_dir()) {
+            return HandleRequest(message.change_dir());
+        } else if (message.has_make_dir()) {
+            return HandleRequest(message.make_dir());
+        } else if (message.has_pwd()) {
+            return HandleRequest(message.pwd());
+        } else if (message.has_quit()) {
+            return HandleRequest(message.quit());
+        } else if (message.has_terminate()) {
+            return HandleRequest(message.terminate());
+        }
+
+        std::cerr << "No valid message was received." << std::endl;
+        return ClientState::ERROR;
+    }
+
+    bool Agent::SendResponse(const Response &response) {
+        // Serialize the message.
+        if (!wire_protocol::Serialize(response, &outgoing_message_buffer_)) {
+            std::cerr << "Failed to serialize message." << std::endl;
+            return false;
+        }
+
+        // Send the message.
+        if (send(client_fd_, outgoing_message_buffer_.data(),
+                 outgoing_message_buffer_.size(), 0) < 0) {
+            perror("Failed to send the message");
+            return false;
+        }
+        return true;
+    }
+
+    bool Agent::SendFileContents(const ftp_messages::FileContents &file_contents, uint16_t command_id) {
+        // Serialize the message.
+        if (!wire_protocol::Serialize(file_contents, &outgoing_message_buffer_)) {
+            std::cerr << "Failed to serialize message." << std::endl;
+            return false;
+        }
+
+        // monitor for termination request for get and put commands
+        uint16_t totalBytesSent = 0;
+
+        // make 1000 byte buffer
+        int numBytesToSend = 1000;
+        uint8_t buf[numBytesToSend];
+
+        // counter used to track
+        int byte_counter = 0;
+
+        bool terminated = false;
+
+        // continue until we've sent the entire message
+        while (totalBytesSent < outgoing_message_buffer_.size()) {
+
+
+            // fill the buffer one byte at a time
+            for (int n:outgoing_message_buffer_) {
```

Comment:
This logic, with all the copying, is kind of clunky. Could you just call `send` repeatedly with the buffer set to `outgoing_message_buffer_.data() + totalBytesSent` and `numBytesToSend`?

Reply:
I'm not exactly sure what you mean by this. I figured there was a better way to go about this but I'm still not super clear on how to do that from that previous comment of yours.

Reply:
Something like:
```c++
while (total_bytes_sent < outgoing_message_buffer_,size()) {
  const uint32_t kBytesToSend = std::min(outgoing_message_buffer.size() - total_bytes_sent, 1000);
  total_bytes_sent += send(client_fd_, outgoing_message_buffer.data() + total_bytes_sent, kBytesToSend);
}
```

Obviously, you would need to add error checking and all the other logic that goes on in this loop, but that's the basic idea.

Reply:
I don't understand why you would want `kBytesToSend`set as the _minimum_ of `outgoing_message_buffer.size() - total_bytes_sent` and 1000? According to the project description, server sends/recieves 1000 bytes at a time. I'm also having trouble understanding what `outgoing_message_buffer.size() - total_bytes_sent` is supposed to represent?